source/core/sweapi/api/Collection.js

```diff
@@ -0,0 +1,123 @@
+/***************************** BEGIN LICENSE BLOCK ***************************
+
+ The contents of this file are subject to the Mozilla Public License, v. 2.0.
+ If a copy of the MPL was not distributed with this file, You can obtain one
+ at http://mozilla.org/MPL/2.0/.
+
+ Software distributed under the License is distributed on an "AS IS" basis,
+ WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
+ for the specific language governing rights and limitations under the License.
+
+ Copyright (C) 2015-2021 Georobotix Inc. All Rights Reserved.
+
+ Author: Mathieu Dhainaut <mathieu.dhainaut@gmail.com>
+
+ ******************************* END LICENSE BLOCK ***************************/
+
+import {isDefined} from "../../utils/Utils";
+
+class Collection {
+    /**
+     *
+     */
+    constructor(url, filter, pageSize, parser, connector) {
+        this.url = url;
+        this.filter = filter;
+        this.pageSize = pageSize;
+        this.parser = parser;
+        this.connector = connector;
+        this.currentOffset = 0;
+        this.nextOffset = undefined;
+        this.previousOffset = undefined;
+        this.init = false;
+        this.total = 0;
+        this.data = [];
+    }
+
+    // cannot fetch dynamically data here otherwise we must change this function into async
+    hasNext() {
+        return !isDefined(this.nextOffset) || (this.nextOffset !== -1);
+    }
+
+    async fetchData(offset) {
+        const response = await this.connector.doRequest(
+            this.url,
+            `${this.filter.toQueryString()}&offset=${offset}&limit=${this.pageSize}`,
+            this.filter.props.responseType
+            );
+        await this.onFetchData(response);
+    }
+
+    async onFetchData(response) {
+        await this.parseResponse(response);
+    }
+
+    async parseResponse(encodedResponse) {
+        let rec = JSON.parse(String.fromCharCode.apply(null, new Uint8Array(encodedResponse)));
+
+        this.data = [];
```

Comment:
I think it would be better to create a new array for each page instead of reusing the same class variable so that previous pages don't get overwritten (the consumer may want to cache the pages)